staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -443,6 +443,12 @@ func (c *Cacher) startCaching(stopChannel <-chan struct{}) {
 	c.watchCache.SetOnReplace(func() {
 		successfulList = true
 		c.ready.set(true)
+		// We rely on the fact that onReplace callback is called
+		// with watchCache lock being held.
+		// That ensures that Watch can't be registered (as it also
+		// grabs the watchcache lock) and ensures no deadlock as
+		// we always grab watchcache lock before grabbing cacher lock.
+		c.terminateAllWatchers()
```

Comment:
I believe this is fixing https://github.com/kubernetes/kubernetes/issues/116122 but I didn't yet manage to reproduce this behavior with test - will try to work on it later this week...

Reply:
if you terminate just after you set to false, new watches have to be waiting for ready true, does it help?

Reply:
there is still a race then - because they may actually already checked the ready condition but didn't yet register themself in all-watchers...

Reply:
OK - so I managed to reproduce it - the test is added below.